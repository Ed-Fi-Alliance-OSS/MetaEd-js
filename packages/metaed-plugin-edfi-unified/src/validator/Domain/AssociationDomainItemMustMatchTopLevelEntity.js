// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'association') {
        if (!metaEd.entity.association.has(domainItem.metaEdName) && !metaEd.entity.associationSubclass.has(domainItem.metaEdName)) {
          failures.push(getFailure(domainItem, 'AssociationDomainItemMustMatchTopLevelEntity',
            `Association Domain Item property '${domainItem.metaEdName}' does not match any declared Association or Association Subclass.`));
        }
      }
    });
  });

  return failures;
}
