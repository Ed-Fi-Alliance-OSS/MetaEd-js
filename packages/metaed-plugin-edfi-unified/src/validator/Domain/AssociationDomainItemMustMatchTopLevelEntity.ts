import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.metaEdName,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'association') return;
        if (
          getEntityFromNamespaceChain(
            domainItem.metaEdName,
            domainItem.referencedNamespaceName,
            namespace,
            'association',
            'associationSubclass',
          ) == null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'AssociationDomainItemMustMatchTopLevelEntity',
              `Association Domain Item property '${
                domainItem.metaEdName
              }' does not match any declared Association or Association Subclass in namespace ${
                domainItem.referencedNamespaceName
              }.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
