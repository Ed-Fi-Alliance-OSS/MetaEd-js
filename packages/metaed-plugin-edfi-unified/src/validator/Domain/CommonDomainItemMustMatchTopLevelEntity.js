// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'common') {
        if (!metaEd.entity.common.has(domainItem.metaEdName)) {
          failures.push(getFailure(domainItem, 'CommonDomainItemMustMatchTopLevelEntity',
            `Common Domain Item property '${domainItem.metaEdName}' does not match any declared Common.`));
        }
      }
    });
  });

  return failures;
}
