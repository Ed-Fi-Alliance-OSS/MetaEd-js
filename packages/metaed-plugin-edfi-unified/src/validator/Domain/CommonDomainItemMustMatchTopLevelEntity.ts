import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.metaEdName,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'common') return;
        if (
          getEntityFromNamespaceChain(domainItem.metaEdName, domainItem.referencedNamespaceName, namespace, 'common') == null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'CommonDomainItemMustMatchTopLevelEntity',
              `Common Domain Item property '${domainItem.metaEdName}' does not match any declared Common in namespace ${
                domainItem.referencedNamespaceName
              }.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
