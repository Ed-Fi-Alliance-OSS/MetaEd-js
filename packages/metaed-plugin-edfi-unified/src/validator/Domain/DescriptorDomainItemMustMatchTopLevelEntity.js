// @flow
import type { DomainItem, MetaEdEnvironment, ValidationFailure } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.type,
    fileMap: null,
  };
}

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    domain.domainItems.forEach(domainItem => {
      if (domainItem.referencedType === 'descriptor') {
        if (!metaEd.entity.descriptor.has(domainItem.metaEdName)) {
          failures.push(
            getFailure(
              domainItem,
              'DescriptorDomainItemMustMatchTopLevelEntity',
              `Descriptor Domain Item property '${domainItem.metaEdName}' does not match any declared Descriptor.`,
            ),
          );
        }
      }
    });
  });

  return failures;
}
