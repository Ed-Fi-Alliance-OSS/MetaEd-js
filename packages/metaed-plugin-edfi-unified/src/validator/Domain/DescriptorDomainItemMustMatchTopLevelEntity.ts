import { DomainItem, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

function getFailure(domainItem: DomainItem, name: string, failureMessage: string): ValidationFailure {
  return {
    validatorName: name,
    category: 'error',
    message: failureMessage,
    sourceMap: domainItem.sourceMap.metaEdName,
    fileMap: null,
  };
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      domain.domainItems.forEach(domainItem => {
        if (domainItem.referencedType !== 'descriptor') return;
        if (
          getEntityFromNamespaceChain(domainItem.metaEdName, domainItem.referencedNamespaceName, namespace, 'descriptor') ==
          null
        ) {
          failures.push(
            getFailure(
              domainItem,
              'DescriptorDomainItemMustMatchTopLevelEntity',
              `Descriptor Domain Item property '${
                domainItem.metaEdName
              }' does not match any declared Descriptor in namespace ${domainItem.referencedNamespaceName}.`,
            ),
          );
        }
      });
    });
  });

  return failures;
}
