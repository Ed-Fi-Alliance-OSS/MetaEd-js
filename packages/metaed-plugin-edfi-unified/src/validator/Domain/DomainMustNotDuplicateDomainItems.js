// @flow
import type { MetaEdEnvironment } from '../../../../../packages/metaed-core/src/MetaEdEnvironment';
import type { ValidationFailure } from '../../../../../packages/metaed-core/src/validator/ValidationFailure';
import { findDuplicates } from '../ValidatorShared/FindDuplicates';

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domain.forEach(domain => {
    const names = domain.domainItems.map((di) => di.metaEdName);

    const duplicates: Array<string> = findDuplicates(names);
    duplicates.forEach(val => {
      const domainItem = domain.domainItems.find((d) => d.metaEdName === val);
      if (domainItem !== undefined) {
        failures.push({
          validatorName: 'DomainMustNotDuplicateDomainItems',
          category: 'error',
          message: `Domain Entity Domain Item '${domainItem.metaEdName}' has a duplicate Domain Item.`,
          sourceMap: domainItem.sourceMap.type,
          fileMap: null,
        });
      }
    });
  });

  return failures;
}
