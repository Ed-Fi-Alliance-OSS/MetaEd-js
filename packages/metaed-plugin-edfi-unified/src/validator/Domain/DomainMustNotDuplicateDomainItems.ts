import { MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { findDuplicates } from '../ValidatorShared/FindDuplicates';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domain.forEach(domain => {
      const names = domain.domainItems.map(di => di.metaEdName);
      const duplicates: Array<string> = findDuplicates(names);

      duplicates.forEach(val => {
        const domainItem = domain.domainItems.find(d => d.metaEdName === val);
        if (domainItem !== undefined) {
          failures.push({
            validatorName: 'DomainMustNotDuplicateDomainItems',
            category: 'error',
            message: `Domain Entity Domain Item '${domainItem.metaEdName}' has a duplicate Domain Item.`,
            sourceMap: domainItem.sourceMap.type,
            fileMap: null,
          });
        }
      });
    });
  });

  return failures;
}
