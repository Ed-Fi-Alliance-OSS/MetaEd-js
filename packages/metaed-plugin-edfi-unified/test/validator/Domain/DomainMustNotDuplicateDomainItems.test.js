// @flow
import { newMetaEdEnvironment, MetaEdTextBuilder, DomainBuilder } from '../../../../../packages/metaed-core/index';
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';
import { validate } from '../../../src/validator/Domain/DomainMustNotDuplicateDomainItems';

describe('when validating domain entity domain item does not duplicate domain items', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName: string = 'DomainName';

  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('namespace', 'ProjectExtension')
      .withStartDomain(domainName, '1')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('DomainItem1')
      .withDomainEntityDomainItem('DomainItem2')
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()
      .sendToListener(new DomainBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one domain entity', () => {
    expect(metaEd.entity.domain.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating domain entity domain item duplicates domain items', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName: string = 'DomainName';
  const domainEntityName: string = 'DomainEntityName';

  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('namespace', 'ProjectExtension')
      .withStartDomain(domainName, '1')
      .withDocumentation('doc')
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName)
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()
      .sendToListener(new DomainBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one domain entity', () => {
    expect(metaEd.entity.domain.size).toBe(1);
  });

  it('should have one validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('DomainMustNotDuplicateDomainItems');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when domain entity domain item has duplicate should have validation failure -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when domain entity domain item has duplicate should have validation failure -> sourceMap');
  });
});

describe('when validating domain entity domain item has multiple duplicate domain items', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainName: string = 'DomainName';
  const domainEntityName: string = 'DomainEntityName';
  const domainEntityName2: string = 'DomainEntityName2';

  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('namespace', 'ProjectExtension')
      .withStartDomain(domainName, '1')
      .withDocumentation('doc')
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName2)
      .withDomainEntityDomainItem(domainEntityName2)
      .withDomainEntityDomainItem('NotDuplicate')
      .withFooterDocumentation('FooterDocumentation')
      .withEndDomain()
      .sendToListener(new DomainBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one domain entity', () => {
    expect(metaEd.entity.domain.size).toBe(1);
  });

  it('should have multiple validation failures', () => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('DomainMustNotDuplicateDomainItems');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when domain entity domain item has duplicate should have validation failure -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when domain entity domain item has duplicate should have validation failure -> sourceMap');
    expect(failures[1].validatorName).toBe('DomainMustNotDuplicateDomainItems');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot('when domain entity domain item has duplicate should have validation failure -> message');
    expect(failures[1].sourceMap).toMatchSnapshot('when domain entity domain item has duplicate should have validation failure -> sourceMap');
  });
});
