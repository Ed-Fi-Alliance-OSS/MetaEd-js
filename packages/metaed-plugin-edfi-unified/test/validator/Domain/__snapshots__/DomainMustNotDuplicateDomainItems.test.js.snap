// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating domain entity domain item duplicates domain items should have one validation failure: when domain entity domain item has duplicate should have validation failure -> message 1`] = `"Domain Entity Domain Item 'DomainEntityName' has a duplicate Domain Item."`;

exports[`when validating domain entity domain item duplicates domain items should have one validation failure: when domain entity domain item has duplicate should have validation failure -> sourceMap 1`] = `undefined`;

exports[`when validating domain entity domain item has multiple duplicate domain items should have multiple validation failures: when domain entity domain item has duplicate should have validation failure -> message 1`] = `"Domain Entity Domain Item 'DomainEntityName' has a duplicate Domain Item."`;

exports[`when validating domain entity domain item has multiple duplicate domain items should have multiple validation failures: when domain entity domain item has duplicate should have validation failure -> message 2`] = `"Domain Entity Domain Item 'DomainEntityName2' has a duplicate Domain Item."`;

exports[`when validating domain entity domain item has multiple duplicate domain items should have multiple validation failures: when domain entity domain item has duplicate should have validation failure -> sourceMap 1`] = `undefined`;

exports[`when validating domain entity domain item has multiple duplicate domain items should have multiple validation failures: when domain entity domain item has duplicate should have validation failure -> sourceMap 2`] = `undefined`;
